lib/config/helper.go

```diff
@@ -0,0 +1,308 @@
+}
+
+func validateStringBasedBoolean(value interface{}, fieldName string) error {
+	switch v := value.(type) {
+	case string:
+		// Check if the string represents a valid boolean value
+		_, err := strconv.ParseBool(v)
+		if err != nil {
+			return fmt.Errorf(fmt.Sprintf(INVALID_FIELD_VALUE, fieldName, value))
+		}
+	default:
+		return fmt.Errorf(fmt.Sprintf("Invalid %s: %v", fieldName, value))
+	}
+	return nil
+}
+
+func validateRequiredPathField(value string, fieldName string) error {
+	if len(value) == 0 {
+		return fmt.Errorf(fmt.Sprintf(INVALID_EMPTY_VALUE, fieldName))
+	}
+	if _, err := os.Stat(value); err != nil {
+		return fmt.Errorf(fmt.Sprintf("Invalid %s: %s (%v)", fieldName, value, err))
+	}
+	return nil
+}
+
+func validateRequiredStringTypeField(value string, fieldName string, expectedValues ...string) error {
+	if len(strings.TrimSpace(value)) < 1 {
+		return fmt.Errorf(fmt.Sprintf(INVALID_EMPTY_VALUE, fieldName))
+	}
+	if len(expectedValues) > 0 {
+		if !contains(expectedValues, value) {
+			expectedValuesStr := strings.Join(expectedValues, ", ")
+			return fmt.Errorf(fmt.Sprintf("Invalid value '%s' for field '%s'. Expected values are: %s", value, fieldName, expectedValuesStr))
+		}
+	}
+	return nil
+}
+
+func contains(slice []string, value string) bool {
+	for _, v := range slice {
+		if v == value {
+			return true
+		}
+	}
+	return false
+}
+
+func validateStringTypeField(value interface{}, fieldName string) error {
+	if value == nil {
+		return nil
+	}
+	_, ok := value.(string)
+	if !ok {
+		return fmt.Errorf(fmt.Sprintf(INVALID_STRING_TYPE, fieldName))
+	}
+	return nil
+}
+
+func validateRequiredStringListField(value []string, fieldName string) error {
+	if len(value) < 1 {
+		return fmt.Errorf(fmt.Sprintf(INVALID_EMPTY_VALUE, fieldName))
+	}
+	return nil
+}
+
+func getSingleErrorFromList(errorList *list.List) error {
+	if errorList.Len() > 0 {
+		var errorMsgs []string
+		for e := errorList.Front(); e != nil; e = e.Next() {
+			switch value := e.Value.(type) {
+			case error:
+				errorMsgs = append(errorMsgs, value.Error())
+			case string:
+				errorMsgs = append(errorMsgs, value)
+			default:
+				errorMsgs = append(errorMsgs, fmt.Sprintf("Unknown error type: %v", value))
+			}
+		}
+		return fmt.Errorf(strings.Join(errorMsgs, "\n"))
+	}
+	return nil
+}
+
+func validateUrl(value string, keyName string) error {
+	if len(value) < 1 {
+		return fmt.Errorf("Invalid or empty URL: " + keyName)
+	}
+	if strings.Contains(value, " ") {
+		return fmt.Errorf("Domain name cannot contain spaces: " + keyName)
+	}
+	// Check for "http://" or "https://" in the URL
+	if strings.HasPrefix(value, "http://") || strings.HasPrefix(value, "https://") {
+		return fmt.Errorf("URL should not include the protocol (http:// or https://): " + keyName)
+	}
+	// Regular expression to validate URLs with or without port number
+	regex := `^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])(:\d+)?$`
```

Comment:
the value `testurl.com` is passed for this regex, which should be a non-valid scenario

Reply:
Updated code.